keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
+
+### Kubernetes Core vs AddOns Component Configs
+
+Currently kubeadm maintains the component configs of both kubelet and kube-proxy. These are components that are maintained in very different ways.
+
+Kubelet, along with API Server, Scheduler, Controller Manager and etcd, are deemed essential Kubernetes components. They are deployed and managed locally by kubeadm, using systemd services (kubelet) or static pods on the local kubelet (API Server, Scheduler, etc.). Hence, their configs are usually consumed by the components in the form of local files.
+
+On the other hand, kube-proxy and CoreDNS/kube-dns are addons, that are managed as standard Kubernetes deployments. Hence their configs are usually accessed via config maps.
+
+Therefore, two different config lifecycles are distinguishable here - one for Kubernetes core components and one for cluster addons.
+
+The defaulting and validation changes, proposed in the previous paragraphs, are going to be applied to both Kubernetes core and AddOn component config types.
+
+#### AddOn Component Config management
+
+AddOn component configs are always stored in config maps. They can be supplied in their entirety by users along with the kubeadm config during init or can be generated by kubeadm itself.
+If, upon upgrade, kubeadm detects an obsolete config version, it will bail out with an error message, asking users to upgrade the addon config by hand.
+
+The ultimate goal for addon component configs, is for them to be eventually managed in their entirety by cluster addon operators, instead of kubeadm.
+
+#### Kubernetes Core Component Config management
+
+Like in the addon case, Kubernetes core component configs can be specified by users in their entirety along with the kubeadm config during init or in a config map. On the other hand, if those configs are not user supplied, they are generated by kubeadm.
+User supplied configs need to be stored in config maps and have to be manually converted by users upon upgrade, while kubeadm generated ones need neither of these things.  
+As most users would choose kubeadm to generate and manage Kubernetes core component configs itself, it is a viable proposition to treat generated and user supplied configs differently.
+
+##### Generated Kubernetes Core Component Config management
+
+Whenever kubeadm does not find component configs, in its config file or in a set of well known config maps, it will opt in for generation.  
+Generated core configs won’t be stored in config maps. Instead, they will always be regenerated in their entirety. This will allow seamless component config version upgrades when users have delegated config management responsibility to kubeadm.
+
+##### User supplied Kubernetes Core Component Config management
```

Comment:
What about adding a few lines about the UX for providing user configs?
The most interesting part here is handling of cluster-wide configs vs node-specific configs; node-specific config as of today are not supported by kubeadm, if possible they will be a nice addition and allow getting read of the kubelet env var file

Reply:
With Kustomize support for Kuberntes Core Component configs, patching the config for the local kubelet using JSON patches will be entirely possible.